tests/reconstructor/test_2D_QLIPP_recon.py

```diff
@@ -0,0 +1,131 @@
+    
+    # initiate simulator
+    simulator = wo.waveorder_microscopy_simulator((N,M), lambda_illu, ps, NA_obj, NA_illu, z_defocus, chi, n_media=n_media,\
+                                                  illu_mode='Arbitrary', Source=Source_discrete)
+
+    I_meas, _ = simulator.simulate_waveorder_measurements(t_eigen, sa, multiprocess=False)
+
+
+    # initiate reconstructor
+    setup = wo.waveorder_microscopy((N,M), lambda_illu, ps, NA_obj, NA_illu, z_defocus, chi, n_media=n_media, 
+                                    phase_deconv='2D', bire_in_plane_deconv='2D', illu_mode='BF')
+
+
+    S_image_recon = setup.Stokes_recon(I_meas)
+    S_image_tm = setup.Stokes_transform(S_image_recon)
+    Recon_para = setup.Polarization_recon(S_image_tm) # Without accounting for diffraction
+
+
+    # Tikhonov regularizer for phase
+    reg_u = 1e-5
+    reg_p = 1e-5
+    S0_stack = S_image_recon[0].copy()
+
+    mu_sample, phi_sample = setup.Phase_recon(S0_stack, method='Tikhonov', reg_u = reg_u, reg_p = reg_p)
+
+
+    # TV regularizer for phase
+    lambda_u = 3e-3
+    lambda_p = 1e-3
+    S0_stack = S_image_recon[0].copy()
+
+    mu_sample_TV, phi_sample_TV = setup.Phase_recon(S0_stack, method='TV', lambda_u = lambda_u, lambda_p = lambda_p, itr = 10, rho=1)
+
+
+    # Diffraction aware reconstruction assuming slowly varying transmission
+    S1_stack = S_image_recon[1].copy()/S_image_recon[0].mean()
+    S2_stack = S_image_recon[2].copy()/S_image_recon[0].mean()
+
+    # Tikhonov
+    retardance, azimuth = setup.Birefringence_recon_2D(S1_stack, S2_stack, method='Tikhonov', reg_br = 1e-3)
+
+    # TV
+    retardance_TV, azimuth_TV = setup.Birefringence_recon_2D(S1_stack, S2_stack, method='TV', reg_br = 1e-1,\
+                                                          rho = 1e-5, lambda_br=1e-3, itr = 20, verbose=True)
+
+
+    # Compute ground truth images
+
+    phase_gt = (phi_s + phi_f)/2
+    phase_gt = np.real(ifft2(np.mean(fft2(phase_gt)[:,:,np.newaxis]*np.abs(setup.Hp)**2 / (np.abs(setup.Hp)**2+reg_p),axis=2)))
+    phase_gt -= np.mean(phase_gt)
+
+    S1_gt = (phi_s - phi_f)*np.sin(2*sa)
+    S2_gt = (phi_s - phi_f)*np.cos(2*sa)
+
+
+    # test the accuracy of reconstruction using the relative error
+
+    # phase_Tikhonov (sensitive to regularization tuning, might cause error if the regularization is not tuned properly)
+    assert(np.sum(np.abs(phase_gt-(phi_sample-np.mean(phi_sample)))**2)/np.sum(np.abs(phase_gt)**2) < 0.1)
+
+    # phase_TV
+    assert(np.sum(np.abs(phase_gt-(phi_sample_TV-np.mean(phi_sample_TV)))**2)/np.sum(np.abs(phase_gt)**2) < 0.1)
+
+    # retardance + orientation (no deconvolution)
+    assert(np.sum(np.abs(S1_gt-Recon_para[0,:,:,2]*np.sin(2*Recon_para[1,:,:,2]))**2)/np.sum(np.abs(S1_gt)**2) < 0.1)
+    assert(np.sum(np.abs(S2_gt-Recon_para[0,:,:,2]*np.cos(2*Recon_para[1,:,:,2]))**2)/np.sum(np.abs(S2_gt)**2) < 0.1)
```

Comment:
When would the `<0.1` condition be relevant?  Only any time you do Decon right?  Could the value be even less here?

Related too -- when you are doing decon, what is a reasonable difference between simulated and reconstructed?

Reply:
Thanks for bringing it up. This is a tricky question. Generally speaking, retardance measurement captures the DC component of the images but the phase measurement does not. Therefore, retardance measurement has overall lower error between the reconstructions and the ground true images. Potentially, we can use 0.5 % here for retardance. Here I get these numbers empirically by actually comparing the results from this current pipeline. Now 10% serves its function to detect major changes (bugs) when there is a major refactoring. We can iterate this next time we work on this test.